grad-cam.py

```diff
@@ -85,59 +73,96 @@ def deprocess_image(x):
+    method for visualizing input saliency.
+    See https://arxiv.org/abs/1412.6806 """
+    input_img = model.input
+    layer_output = model.get_layer(activation_layer).output
+    grads = K.gradients(layer_output, input_img)[0]
+    gradient_fn = K.function([input_img, K.learning_phase()], [grads])
+    grads_val = gradient_fn([img, 0])[0]
+    return grads_val
 
-preprocessed_input = load_image(sys.argv[1])
 
-model = VGG16(weights='imagenet')
+def grad_cam(input_model, img, category_index, activation_layer):
+    """GradCAM method for visualizing input saliency."""
+    loss = input_model.output[0, category_index]
+    layer_output = input_model.get_layer(activation_layer).output
+    grads = normalize(K.gradients(loss, layer_output)[0])
+    gradient_fn = K.function([input_model.input, K.learning_phase()], [layer_output, grads])
 
-predictions = model.predict(preprocessed_input)
-top_1 = decode_predictions(predictions)[0][0]
-print('Predicted class:')
-print('%s (%s) with probability %.2f' % (top_1[1], top_1[0], top_1[2]))
+    conv_output, grads_val = gradient_fn([img, 0])
+    conv_output, grads_val = conv_output[0], grads_val[0]
 
-predicted_class = np.argmax(predictions)
-cam, heatmap = grad_cam(model, preprocessed_input, predicted_class, "block5_conv3")
-cv2.imwrite("gradcam.jpg", cam)
+    weights = np.mean(grads_val, axis=(0, 1))
+    cam = np.dot(conv_output, weights)
 
-register_gradient()
-guided_model = modify_backprop(model, 'GuidedBackProp')
-saliency_fn = compile_saliency_function(guided_model)
-saliency = saliency_fn([preprocessed_input, 0])
-gradcam = saliency[0] * heatmap[..., np.newaxis]
-cv2.imwrite("guided_gradcam.jpg", deprocess_image(gradcam))
+    cam = cv2.resize(cam, (224, 224), cv2.INTER_LINEAR)
+    cam = np.maximum(cam, 0)
+    cam = cam / cam.max()
+    return cam
+
+
+def compute_saliency(model, guided_model, layer_name, img_path, cls=-1, visualize=True, save=True):
+    preprocessed_input = load_image(img_path)
+
+    predictions = model.predict(preprocessed_input)
+    top_n = 5
+    top = decode_predictions(predictions, top=top_n)[0]
+    classes = np.argsort(predictions[0])[-top_n:][::-1]
+    print('Model prediction:')
+    for c, p in zip(classes, top):
+        print('\t({}) {:20s}\twith probability {:.3f}'.format(c, p[1],p[2]))
+    if cls == -1:
+        cls = np.argmax(predictions)
+    nb_classes = 1000
+    class_name = decode_predictions(np.eye(1, nb_classes, cls))[0][0][1]
+    print("Computing saliency for '{}'".format(class_name))
+
+    gradcam = grad_cam(model, preprocessed_input, cls, activation_layer=layer_name)
+    gb = guided_backprop(guided_model, img=preprocessed_input, activation_layer=layer_name)
+    guided_gradcam = gb * gradcam[..., np.newaxis]
+
+    if save:
+        jetcam = cv2.applyColorMap(np.uint8(255 * gradcam), cv2.COLORMAP_JET)
+        jetcam = (np.float32(jetcam) + np.float32(cv2.imread(sys.argv[1]))) / 2
```

Comment:
I would put the arguments of `np.float32( cv2.imread (...) )` loading of the image as  `np.float32(image.load_img(sys.argv[1], target_size=(224, 224)))` instead.

Otherwise, it would break if the image has a different size when being loaded by the `load_image` function.
In best case scenario, the size of the image needs to be taken out as a separate parameter - e.g. top of the file.